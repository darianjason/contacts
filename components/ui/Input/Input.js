import React, {useReducer, useEffect} from 'react';
import {View, Text, TextInput} from 'react-native';

import {INPUT_CHANGE, INPUT_BLUR} from '../../../constants';
import DefaultText from '../DefaultText/DefaultText';
import styles from './Input.styles';

const inputReducer = (state, action) => {
  switch (action.type) {
    case INPUT_CHANGE:
      return {
        ...state,
        value: action.value,
        isValid: action.isValid,
        errorText: action.errorText,
      };
    case INPUT_BLUR:
      return {
        ...state,
        isTouched: true,
      };
    default:
      return state;
  }
};

const validateRequired = (value, required, validation) => {
  if (required && !value.trim()) {
    validation.isValid = false;
    validation.errorText = 'Required';
  }
};

const validateType = (value, alphanumeric, numeric, validation) => {
  const alphanumericRegex = /^[a-z0-9]+$/i;

  if (alphanumeric && !alphanumericRegex.test(value)) {
    validation.isValid = false;
    validation.errorText = 'Must be alphanumeric';
  } else if (numeric && isNaN(+value)) {
    validation.isValid = false;
    validation.errorText = 'Must be numeric';
  }
};

const validateAmount = (value, min, max, validation) => {
  if (min && +value < min) {
    validation.isValid = false;
    validation.errorText = `Must be larger than ${min - 1}`;
  } else if (max && +value > max) {
    validation.isValid = false;
    validation.errorText = `Must be smaller than ${max + 1}`;
  }
};

const validateLength = (value, minLength, maxLength, validation) => {
  if (minLength && value.length < minLength) {
    validation.isValid = false;
    validation.errorText = `Must be longer than ${minLength - 1} characters`;
  } else if (maxLength && value.length > maxLength) {
    validation.isValid = false;
    validation.errorText = `Must be shorter than ${maxLength + 1} characters`;
  }
};

export const validate = (value, validationParams, validation) => {
  const {
    required, min, max, minLength,
    maxLength, alphanumeric, numeric
  } = validationParams;

  validateLength(value, minLength, maxLength, validation);
  validateAmount(value, min, max, validation);
  validateType(value, alphanumeric, numeric, validation);
  validateRequired(value, required, validation);
};

const textChangeHandler = (value, validationParams, dispatch) => {
  const validation = {
    isValid: true,
    errorText: '',
  };

  validate(value, validationParams, validation);

  dispatch({
    type: INPUT_CHANGE,
    value: value,
    isValid: validation.isValid,
    errorText: validation.errorText,
  });
};

const loseFocusHandler = dispatch => {
  dispatch({type: INPUT_BLUR});
};

const Input = props => {
  const {
    id,
    style,
    label,
    initialValue,
    initiallyValid,
    required,
    min,
    max,
    minLength,
    maxLength,
    alphanumeric,
    numeric,
    onInputChange,
  } = props;

  const [inputState, dispatch] = useReducer(inputReducer, {
    value: initialValue ? initialValue : '',
    isValid: initiallyValid,
    isTouched: false,
    errorText: '',
  });

  const {value, isValid, isTouched, errorText} = inputState;

  useEffect(() => {
    onInputChange(id, value, isValid, errorText);
  }, [onInputChange, id, value, isValid, errorText]);

  const validationParams = {
    required,
    min,
    max,
    minLength,
    maxLength,
    alphanumeric,
    numeric,
  };

  return (
    <View style={[styles.container, style]}>
      <Text
        style={[styles.label, !isValid && isTouched && styles.invalidLabel]}
      >
        {label}
      </Text>
      <TextInput
        {...props}
        style={[styles.input, !isValid && isTouched && styles.invalidInput]}
        value={value}
        onChangeText={text =>
          textChangeHandler(text, validationParams, dispatch)
        }
        onBlur={() => loseFocusHandler(dispatch)}
      />
      {!isValid && isTouched && (
        <DefaultText style={styles.errorText}>{errorText}</DefaultText>
      )}
    </View>
  );
};

export default Input;
