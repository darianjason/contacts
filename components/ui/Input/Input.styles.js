import {StyleSheet} from 'react-native';

import {Colors} from '../../../constants';

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
    justifyContent: 'center',
  },
  label: {
    fontFamily: 'Manrope-Bold',
    letterSpacing: 0.5,
    color: Colors.primary,
    marginBottom: 8,
  },
  input: {
    fontFamily: 'Manrope-Regular',
    paddingHorizontal: 8,
    paddingVertical: 4,
    borderRadius: 4,
    backgroundColor: Colors.white,
  },
  invalidLabel: {
    color: Colors.red,
  },
  invalidInput: {
    borderWidth: 2,
    borderStyle: 'solid',
    borderColor: Colors.red,
  },
  errorText: {
    marginTop: 4,
    fontSize: 12,
    color: Colors.red,
  },
});

export default styles;
