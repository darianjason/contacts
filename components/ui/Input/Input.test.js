import React from 'react';
import renderer from 'react-test-renderer';

import Input, {validate} from './Input';

describe('Input component', () => {
  const validation = {
    isValid: false,
    errorText: '',
  };

  test('invalid and error text is "Required" if required is set but no value', () => {
    const value = '';
    const validationParams = {required: true};

    validate(value, validationParams, validation);

    expect(validation.isValid).toBe(false);
    expect(validation.errorText).toBe('Required');
  });

  test('invalid and error text is "Must be alphanumeric" if alphanumeric is set but input is not alphanumeric', () => {
    const value = '!@#';
    const validationParams = {alphanumeric: true};

    validate(value, validationParams, validation);

    expect(validation.isValid).toBe(false);
    expect(validation.errorText).toBe('Must be alphanumeric');
  });

  test('invalid and error text is "Must be larger than ..." if minimum value is set but input is smaller', () => {
    const value = 5;
    const validationParams = {min: 10};

    validate(value, validationParams, validation);

    expect(validation.isValid).toBe(false);
    expect(validation.errorText).toBe('Must be larger than 9');
  });

  test('invalid and error text is "Must be smaller than ..." if maximum value is set but input is larger', () => {
    const value = 2;
    const validationParams = {max: 1};

    validate(value, validationParams, validation);

    expect(validation.isValid).toBe(false);
    expect(validation.errorText).toBe('Must be smaller than 2');
  });

  test('invalid and error text is "Must be longer than ... characters" if minimum length is set but input is shorter', () => {
    const value = 'asd';
    const validationParams = {minLength: 5};

    validate(value, validationParams, validation);

    expect(validation.isValid).toBe(false);
    expect(validation.errorText).toBe('Must be longer than 4 characters');
  });

  test('invalid and error text is "Must be shorter than ... characters" if maximum length is set but input is longer', () => {
    const value = 'hello';
    const validationParams = {maxLength: 3};

    validate(value, validationParams, validation);

    expect(validation.isValid).toBe(false);
    expect(validation.errorText).toBe('Must be shorter than 4 characters');
  });

  test('renders correctly', () => {
    const tree = renderer.create(<Input onInputChange={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
