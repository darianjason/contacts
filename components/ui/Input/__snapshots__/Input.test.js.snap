// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "justifyContent": "center",
        "marginBottom": 16,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#142B3E",
          "fontFamily": "Manrope-Bold",
          "letterSpacing": 0.5,
          "marginBottom": 8,
        },
        false,
      ]
    }
  />
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    onInputChange={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#F7F4F3",
          "borderRadius": 4,
          "fontFamily": "Manrope-Regular",
          "paddingHorizontal": 8,
          "paddingVertical": 4,
        },
        false,
      ]
    }
    value=""
  />
</View>
`;
